### ISSUE TITLE
Parser does fail for a correct region as defined by parantheses if a syntax error is before

### ISSUE THREAD

**Issue by afischli** [supercollider/supercollider#1891]
Here the reproducer for the bug in the parser:

(
1+1; // put the cursor into this line and press Cmd-Return => you get 3, press Shift-Return => you get 2 (All as expected)
1+2;
)

{ Saw.ar(LFNoise1.kr(1).range(100,[200,300]) }.play;

Above code causes no problems and all works as expected. However, if you reverse the ordering to this:

{ Saw.ar(LFNoise1.kr(1).range(100,[200,300]) }.play;

(
1+1; // put the cursor into this line and press Cmd-Return => you get 2 but should get 3, press Shift-Return => you get 2
1+2;
)

You no longer get what you expect, i.e. the parser fails to execute the correct statements within the region as defined by the parantheses. The reason is only the presence of a mistake before/above the region instead of after/below the region. The mistake is in the Saw.ar line, BTW a line taken from the Multichannel-Expansion guide (but now fixed cf. b54e617). But the parser should never get there in the first place, since that faulty line is outside the two parantheses and evaluation should stop at the left paranthesis and not encounter the line of code containing Saw.ar.

I believe the parser should always evaluate correctly everything within a region as defined by parantheses regardless of what is outside of the region.

Note, this issue is a follow up of issue #1786 and has also been partly discussed under #1781, both now closed.

BTW, my guess is also that this bug may explain help browser code execution problems as described in issue #1676. But I have not studied that issue in details. It merely appears to me to be logical that if the parser is context dependent – as above reproducer demonstrates – that parsing may easily fail if a region is embedded in freely written text as is the case with help/tutorial texts.

I believe this issue has nothing to do with a particular system. I have been able to reproduce the bug on several systems and using different versions of SC. The last test was done using SuperCollider 3.7.0-beta1 Built from branch 'HEAD' 9899190 under OS X 10.9.5 (Mavericks).

**Comment by telephon** [comment]
Yes, indeed, this is wrong. Actually it isn't the parser, but the bracket-matching algorithm of the IDE. As *region* should count only something between parentheses at the beginning of a line. Somehow, this detection algorithm fails when it encounters an inconsistency. Then instead of either ignoring it (because what matters is only the currently found region) or reporting the syntax error (so one can fix it).

I'd prefer if regions were treated as locally as possible.

**Comment by jamshark70** [comment]
(heh heh) Searching through code is the fun part.

It seems to be `QTextCursor ScCodeEditor::regionAroundCursor()` in editors/sc-ide/widgets/code_editor/sc_editor.cpp. If I'm reading it right, it starts by tokenizing the whole document (`TokenIterator::rightOf( textDocument()->begin(), 0 )`) and then iterating over top-level bracket pairs until it finds one containing the current cursor position. If the top-level pair is a region (`bracketPairDefinesRegion(bracketPair)`), then it returns that region. There is no recursion, so it doesn't check any bracket pairs within other pairs.

Mismatched brackets mean that the tokenizer contains faulty data. That would explain the failure.

I suppose an alternate algorithm would be to scan outward from the cursor position toward matching pairs repeatedly until you can go no further, and then do the `bracketPairDefinesRegion()` test. sc_editor.cpp also has:

```
void ScCodeEditor::selectBlockAroundCursor()
{
    QTextCursor block_cursor = blockAroundCursor(textCursor());
    if (!block_cursor.isNull())
        setTextCursor(block_cursor);
}
```

`blockAroundCursor()` might be the trick.

**Comment by jamshark70** [comment]
For example, with this alternate, cmd-shift-R *does* properly select the region in the above test case, even with the syntax error before the region. (Without the change, cmd-shift-R fails.)

It does not check that the outermost matching brackets are, in fact, a region -- because I don't know how the TokenIterator works and I don't have time today to figure it out.

```
QTextCursor ScCodeEditor::regionAroundCursor(const QTextCursor & cursor)
{
    QTextCursor block_cursor = textCursor();
    QTextCursor temp_cursor = blockAroundCursor(block_cursor);
    while (!temp_cursor.isNull()) {
        block_cursor = temp_cursor;
        temp_cursor = blockAroundCursor(block_cursor);
    }
    // HERE: get bracket pair from TokenIterator and test.
    // I don't know how to do that.
    return block_cursor;
}
```

**Comment by sofakid** [comment]
If someone is going to work on the bracket matching code, I'd like to point out another issue:

Brackets are matched and indentation is modified, within multiline strings.

Like, you type:

    "
        )
    "

and the string, which should not be modified, is modified to:

    "
    )
    "

**Comment by afischli** [comment]
@jamshark70  
> I suppose an alternate algorithm would be to scan outward from the cursor position toward matching pairs repeatedly until you can go no further, and then do the bracketPairDefinesRegion() test. <

Correct. That is exactly what must be done by any evaluation triggered by command "Evaluate Selection, Line or Region" (Cmd-Enter on a Mac) and not behave similar to an "Evaluate File" command that stops at the next closing parantheses as seen from the current cursor position. 

The scanner, I understand you call this tokenizer, must not work from top towards bottom with an interpretive language such as SC. It must first properly find the region, and only then evaluate it. A simple approach to finding the region is to search forward and backward for parentheses/curly braces pairs from the current cursor position until region begin and region end parentheses or end of file are encountered. The current algorithm scanning from the top is by definition for sure doomed to fail sooner or later. The only difficulty with the needed approach are the possible presence of parentheses or curly braces in comments, which ought all to be ignored. Unfortunately SC follows the bad habit of all C influenced languages of not defining the end of a comment, leading here to a great difficulty for the scanner. Otherwise that algorithm would be a piece of cake. But that is too late now, sic. Consequently the scanner while scanning backwards has to do rather difficult things such as suspecting at every end of a line that a comment may have started and then decide later whether the encountered parentheses are to be ignored or not. What is therefore needed is to work in a buffer and delete all comments line by line and scan for the matching parentheses in the buffer without any comments. Such an algorithm should never fail, regardless of the context and perhaps all the in help evaluations may suddenly start working also robustly. 

While I was working through the "Getting Started Tutorial" I had many examples, some deliberately faulty as introduced by the tutorial, above the code I was currently working on, assuming of course that this would not matter. The result: Terrible confusion and a very bad impression of SC (#1781), since very little worked as described. No wonder! I could then gradually narrow it down to #1786 and finally to this isssue #1891. But that was astonishingly arduous, I wish to remind us all.

**Comment by afischli** [comment]
BTW, is there anywhere a formal definition such as an EBNF of the SC language? I realized I would at least need also a list of its symbols, but can't find it anywhere. Does any of those exist at all?

**Comment by sofakid** [comment]
I've been working on that for my own project.

The parser is a bison one (lang/LangSource/Bison/lang11d), the lexer
appears to be hard-coded.

This is an incomplete conversion from the bison grammar
(lang/LangSource/Bison/lang11d), and a cursory attempt at naming the tokens.

You'll see i've replaced the tokens with CamelCase names, the ones i
haven't yet, are still in CAPS, as well a couple of bison keywords, namely
%prec have been left in there.

https://gist.github.com/sofakid/6ca32911f4edcd76b00e

hope this helps

**Comment by telephon** [comment]
>A simple approach to finding the region is to search forward and backward for parentheses/curly braces pairs from the current cursor position until region begin and region end parentheses or end of file are encountered. 

I agree. We had a discussion here on this earlier, but at that time, things were decided differently. The crucial question that needs to be decided here as well is:

**Comment by telephon** [comment]
>The only difficulty with the needed approach are the possible presence of parentheses or curly braces in comments, which ought all to be ignored. Unfortunately SC follows the bad habit of all C influenced languages of not defining the end of a comment, leading here to a great difficulty for the scanner. 

Can you explain? Normally, the end of a comment is defined as `*/`, and if you are missing it, you get a ```Open ended comment ... started on line 2 in file 'selected text'```

**Comment by jamshark70** [comment]
> Unfortunately SC follows the bad habit of all C influenced languages of not defining the end of a comment, leading here to a great difficulty for the scanner.

Huh? There are two syntaxes for comments:

- Opening delimiter `/*`, closing `*'`
- Opening `//`, closing `\n` -- there is a very clear ending for this. It happens to be an invisible character, but this is rather different from "not define[d]."

**Comment by afischli** [comment]
Any formal language divides any program text into symbols. You call that token, which is not really correct, since token can be any piece of the program text, e.g. a string or part of a number, and corresponds not necessary to a formally well defined symbol. But for the sake of this discussion, you may assume it is the same.

What you state is exactly the problem. `\n` is not an end of comment symbol, since `\n` can also be the end of a line (EOL) or simply white space (BTW, I prefer EOL, since `\n` may not be the actual EOL on some systems, BTW an EOL is a token that can be more than one character). The interpretation of the very same piece of program text becomes then context dependent. Actually in a well defined language EOL would best be simply treated as white space that can be ignored and allows the programmer to structure the program well and freely. 

The alleged convenience for the user not to have to use a specific symbol for characterizing the end of a // comment by merely hitting return results consequently in significant problems for scanning and parsing. As a consequence of such context dependency it may be very difficult if not impossible to write correct programs for such languages. Languages not formally well defined tend to be context dependent. E.g. shell scripts, can never be programmed correctly and you can always find some use in which they fail, since their correct functioning of its statements depends not only on the context within the script but even on the context in which the script is used. And the reason for the failure is merely the impossibility to scan them correctly under all circumstances. What works in one context fails in the other and vice versa. Thus the programmer can never find a robust solution that works always.

I do not understand SC well enough to dare to comment much on its syntax and semantics, but what we discuss here is the fact that there is a 2nd., alternative comment allowed that ends with EOL, instead of a closing symbol as the other form with the two symbols  '`/*`' and  '`*/`' nicely does.

To come back to our problem of backward traversing the sequence of symbols within a region: The scanner (which you may call the tokenizer) has to recognise the symbols to find the begin of the region. Upon encountering an EOL it has no idea what that is, i.e. whether it ends a comment or a code of line. Consequently while traversing further and encountering a '`(`' or '`{`' or '`)`' or '`}`' symbol it does not know whether these ought to be ignored, since being part of a comment or whether they are part of code.

Can anyone tell me what the begin of region symbol exactly is?  May it contain some white space or not? What makes it unique? Is it '`EOL(`' or '`EOL(EOL`' or '`(EOL`'.  Of course the very last would again create a problem, since that symbol could als be present in a // comment. Having experimented with current version of SC I'd say it is '`EOL('` (white space correctly ignored). Correct?

If that is correct, then the solution is very simple. You traverse backward from the current cursor position until you encounter the begin of region symbol, i.e. '`EOL(`' or the begin of the file (any white space ignored).  Of course you do the same forward to the end of region symbol, which is?

I tested this and found that the end of region symbol is unfortunately not well defined. It seems to be just '`)`', since it tolerates white space before and after. Thus it is not at all a unique symbol. A pity. I would have preferred it to be '`EOL)`' while allowing for as much whitespace before or after. 

I understand now that the scanning must be very convoluted, since you need to find the begin of region symbol '`EOL(`' and then find the matching '`)`' for the parantheses in '`EOL(`' while ignoring any parentheses or curly braces in // comments. Unnecessary complicated.

How about defining the end of region symbol to be from now on '`EOL)`'?  

Too late? Causing too many codes to fail? The advantage would be, to be able to have a simple and robust scanner algorithm.

**Comment by afischli** [comment]
P.S.:  I forgot to explicitly state that what I propose would allow for an algorithm that can deal with comments elegantly. // comments don't matter, since they happen to not conflict with the '`EOL(`' and '`EOL)`' symbols. And the other comment form has fortunately a termination symbol, i.e. '`*/`', allowing to skip all between '`*/`' and '`/*`', including of course any begin or end of region symbols, i.e.  '`EOL(`' or '`EOL)`'.

**Comment by jamshark70** [comment]
Ill-defined or no, the fact remains that the IDE's TokenIterator does already handle comments correctly. Whether it is or isn't difficult to scan: Tim and Jakob already did that work for us.

It is probably better to use the TokenIterator, rather than reinvent the wheel or alter a couple of decades of SC convention.

**Comment by telephon** [comment]
@afischli: thank you for your detailed analysis, your observations are correct as far as regions go. The parser and the region detection are separate things, however. The sclang parser knows no regions, this is a much newer feature of the editor. So what you describe is a possible stumbling block indeed, but not of sclang, but for editors.

Regions should just be whatever seems reasonable in everyday use. I guess it is ok to just ignore lines that either begin with `//` or are enclosed within `/*` and `*/`, and *then* check for something like `EOL)` (or a reasonable alternatives you mention). Have you had a look at `TokenIterator`? It should cover it.

What I feel is debatable is that when automatic reformatting is applied, the content of comments isn't left alone. But this is another, related, issue.

**Comment by prko** [comment]
I often put an indent using "shift+tab' or "spaces" to indicate main codes and supplemental codes as the examples below.
Ex. 2 and Ex. 3 do not work.
I think it is not syntax errors.
```supercollider
/* ex. 1: main code */
(
"case 1".postln
)

/* ex. 2:  supplementary code */
  ( // this line begins with two white spaces.
  "case 2".postln  // this line begins with two white spaces.
  ) // this line begins with two white spaces.

/* ex. 3:  supplementary code */
 	( // this line begins with a shft+tab.
	"case 3".postln // this line begins with a shft+tab.
	) // this line begins with a shft+tab.
```

**Comment by telephon** [comment]
when you do automatic indentation on this, the formatting will be all gone.

**Comment by prko** [comment]
Yes...
Thus I do not use automatic indentation when writing codes like ex. 2 and ex. 3....

**Comment by brianlheim** [comment]
> Can anyone tell me what the begin of region symbol exactly is?

The documentation ("SuperCollider IDE § Code Editor") says this:

> Regions make it quick and easy to evaluate larger portions of code. A region is surrounded by parentheses; to be interpreted as a region, the parenthesis must each be the only character on a line.

But that doesn't seem to be entirely true. Rather, it only needs to be the first character on the line AFAICT.

A lot of strong opinions in this thread. I think this is probably the wrong place to be looking for robust behavior. The concept of a region is really only defined for convenience. Like @telephon said above, "whatever seems reasonable in everyday use".

I would also like to note that under the current implementation, this case works fine, and is an important case to consider for any proposed change to this behavior:

```supercollider
(
sqrt
(
5
)
)
```

@afischli we cannot use `EOL(` as a way of identifying the starts of regions; that wouldn't allow for a document where a region begins with an open paren on the first line.

Can we come up with a more rigorous definition of region? I would propose this:

---

A `region` is the code inside a pair of parentheses; each parenthesis is the first character in its line, and neither parenthesis is inside a comment. The parentheses are fully contained either inside the entire file ("top-level") or inside a larger enclosing region. The opening paren is preceded (not counting whitespace) by either the start of the file, the end of another region, or the start of the enclosing region. The closing paren is preceded (not counting whitespace) by either the end of the file, the end of another region, or the end of the enclosing region.

---

Once we decide on a formal definition it will be much easier to agree on a behavior for incorrect cases.

**Comment by htor** [comment]
my five cents: i wrote the code for region/line evaluation in the help browser in javascript. my naïve implemenation starts searching left and right from the cursor position. for region eval, it recursively collects pairs of parens and selects the outermost pair for evaluation. check out https://github.com/supercollider/supercollider/blob/develop/HelpSource/editor.js#L68.
